io/zenoh-links/zenoh-link-shm/src/unicast.rs

```diff
@@ -0,0 +1,439 @@
+    }
+
+    async fn stop_listening(self) {
+        self.listening_task_handle.cancel().await;
+    }
+}
+
+struct UnicastPipeClient;
+impl UnicastPipeClient {
+    async fn connect_to(endpoint: EndPoint) -> ZResult<UnicastPipe> {
+        let (path_uplink, path_downlink, access_mode) = parse_pipe_endpoint(&endpoint);
+
+        // create Locators
+        let local = Locator::new(
+            endpoint.protocol(),
+            path_downlink.as_str(),
+            endpoint.metadata(),
+        )?;
+        let remote = Locator::new(
+            endpoint.protocol(),
+            path_uplink.as_str(),
+            endpoint.metadata(),
+        )?;
+
+        // create uplink
+        let mut uplink = PipeW::new(&path_uplink).await?;
+        // create downlink
+        let mut downlink = PipeR::new(&path_downlink, access_mode).await?;
+
+        // write invitation
+        uplink.write_all(PIPE_INVITATION).await?;
+
+        // read responce
+        let mut responce: [u8; 4] = [0, 0, 0, 0];
+        downlink.read_exact(&mut responce).await?;
+        // check responce
+        if responce != PIPE_INVITATION {
+            bail!("Unexpected responce received during pipe handshake!")
+        }
+
+        Ok(UnicastPipe {
+            r: async_std::sync::RwLock::new(downlink),
+            w: async_std::sync::RwLock::new(uplink),
+            local,
+            remote,
+        })
+    }
+}
+
+struct UnicastPipe {
+    r: async_std::sync::RwLock<PipeR>,
+    w: async_std::sync::RwLock<PipeW>,
+    local: Locator,
+    remote: Locator,
+}
+impl Drop for UnicastPipe {
+    fn drop(&mut self) {}
+}
+#[async_trait]
+impl LinkUnicastTrait for UnicastPipe {
+    async fn close(&self) -> ZResult<()> {
+        log::trace!("Closing SHM Pipe link: {}", self);
+        Ok(())
+    }
+
+    async fn write(&self, buffer: &[u8]) -> ZResult<usize> {
+        //todo: I completely don't like the idea of locking reader\writer each time, but LinkUnicastTrait requires self.r and self.w to be Sync
+        self.w
+            .write()
+            .await
+            .write(buffer)
+            .await
+    }
+
+    async fn write_all(&self, buffer: &[u8]) -> ZResult<()> {
+        //todo: I completely don't like the idea of locking reader\writer each time, but LinkUnicastTrait requires self.r and self.w to be Sync
+        self.w
+            .write()
+            .await
+            .write_all(buffer)
+            .await
+    }
+
+    async fn read(&self, buffer: &mut [u8]) -> ZResult<usize> {
+        //todo: I completely don't like the idea of locking reader\writer each time, but LinkUnicastTrait requires self.r and self.w to be Sync
+        self.r
+            .write()
+            .await
+            .read(buffer)
+            .await
+    }
+
+    async fn read_exact(&self, buffer: &mut [u8]) -> ZResult<()> {
+        //todo: I completely don't like the idea of locking reader\writer each time, but LinkUnicastTrait requires self.r and self.w to be Sync
```

Comment:
By construction in the rest of zenoh logic there is only one task reading at the time. So, an `UnsafeCell` could also do the trick.